execute/tokendata/usdc/http.go

```diff
@@ -0,0 +1,198 @@
+)
+
+type HTTPClient interface {
+	// Get calls the USDC attestation API with the given USDC message hash.
+	// The attestation service rate limit is 10 requests per second. If you exceed 10 requests
+	// per second, the service blocks all API requests for the next 5 minutes and returns an
+	// HTTP 429 response.
+	//
+	// Documentation:
+	//
+	//	https://developers.circle.com/stablecoins/reference/getattestation
+	//	https://developers.circle.com/stablecoins/docs/transfer-usdc-on-testnet-from-ethereum-to-avalanche
+	Get(ctx context.Context, messageHash [32]byte) ([]byte, int, error)
+}
+
+// httpClient is a client for the USDC attestation API. It encapsulates all the details specific to the Attestation API:
+// - rate limiting
+// - cool down period
+// - parsing JSON response and handling errors
+// Therefore AttestationClient is a higher level abstraction that uses httpClient to fetch attestations and can be more
+// oriented around caching/processing the attestation data instead of handling the API specifics.
+type httpClient struct {
+	apiURL     string
+	apiTimeout time.Duration
+	rate       *rate.Limiter
+	// coolDownUntil defines whether requests are blocked or not.
+	coolDownUntil time.Time
+	coolDownMu    *sync.RWMutex
+}
+
+//nolint:revive
+func NewHTTPClient(api string, apiInterval time.Duration, apiTimeout time.Duration) *httpClient {
+	return &httpClient{
+		apiURL:     api,
+		apiTimeout: apiTimeout,
+		rate:       rate.NewLimiter(rate.Every(apiInterval), 1),
+		coolDownMu: &sync.RWMutex{},
+	}
+}
+
+type attestationStatus string
+
+const (
+	attestationStatusSuccess attestationStatus = "complete"
+	attestationStatusPending attestationStatus = "pending_confirmations"
+)
+
+type httpResponse struct {
+	Status      attestationStatus `json:"status"`
+	Attestation string            `json:"attestation"`
+	Error       string            `json:"error"`
+}
+
+func (r httpResponse) validate() error {
+	if r.Error != "" {
+		return fmt.Errorf("attestation API error: %s", r.Error)
+	}
+	if r.Status == "" {
+		return fmt.Errorf("invalid attestation response")
+	}
+	if r.Status != attestationStatusSuccess {
+		return ErrNotReady
+	}
+	return nil
+}
+
+func (r httpResponse) attestationToBytes() ([]byte, error) {
+	attestationBytes, err := hex.DecodeString(strings.TrimPrefix(r.Attestation, "0x"))
+	if err != nil {
+		return nil, err
+	}
+	return attestationBytes, nil
+}
+
+func (h *httpClient) Get(ctx context.Context, messageHash [32]byte) ([]byte, int, error) {
+	var empty []byte
+	// Terminate immediately when rate limited
+	if h.inCoolDownPeriod() {
+		return empty, http.StatusTooManyRequests, ErrRateLimit
+	}
+
+	if h.rate != nil {
+		// Wait blocks until it the attestation API can be called or the
+		// context is Done.
+		if waitErr := h.rate.Wait(ctx); waitErr != nil {
+			return empty, http.StatusTooManyRequests, ErrRateLimit
+		}
+	}
+
+	timeoutCtx, cancel := context.WithTimeoutCause(ctx, h.apiTimeout, ErrTimeout)
```

Comment:
q: if ctx is canceled before the timeout, does timeoutCtx cancel?

Reply:
have you checked the tests? :P yeah, it cancels